docs: tut/pos (#1583)
reach-sh/reach-lang: docs/src/tut/pos/index.md

> # {#point-of-sale} Point-of-sale

Reviewer: ```suggestion
# {#point-of-sale} Point of Sale
```

This tutorial will walk you through creating a business point-of-sale machine, like Toast or Square.

It assumes that you are comfortable with writing simple Reach DApps and have already completed one or two of our [tutorials](##tuts). 

> Complete them all to become a Reach rockstar.

Reviewer: What is the "them" this is referring to?

Author: The tutorials -- edited to make language more clear

### Here is the general flow of our DApp:
1. Admin gives parameters and deploys contract.
2. Admin deposits non-network Loyalty Tokens into contract.
3. Contract exposes API functions.
4. Users purchase items in our store.
5. Users are offered a refund window.
6. Close access to API functions when Loyalty Token balance is zero.
7. Transfer network tokens to Admin.
8. Terminate contract.

## Where the users at?
At this point of designing our Reach DApp, we need to ask ourselves how many and what types of users will be interacting with it.

We know that we need at least one `{!rsh} Participant` to deploy the contract and serve as our `Admin`. We can consider them our shop owner.

:::note
All Reach DApps require at least one `{!rsh} Participant`, because `{!rsh} API` members *do not* have the ability to deploy contracts.
:::

The rest of our users will be customers in our shop, so they are best represented as `{!rsh} API`. This will allow an unlimited number of unknown users to interact with our contract `{!rsh} API` member functions.

Let's write some Reach code for our users.
```
load: /examples/point-of-sale/index.rsh
md5:  2d9202924ee606ab47726038fbde05a3
range: 1-12
```

> - Line 4 is a new feature with Reach 0.1.13, read more about `{!rsh} ALGOExitMode`.

Reviewer: ```suggestion
- Line 4 is a new feature with Reach 0.1.13; we need it because we're using boxes, you can read more about it by clicking `{!rsh} ALGOExitMode`.
```

Now for our `{!rsh} API` users.
```
load: /examples/point-of-sale/index.rsh
md5:  2d9202924ee606ab47726038fbde05a3
range: 13-17
```
- Line 13 specifies our `Buyer` as an `{!rsh} API`.
- Line 14-15 declare the functions that this type of user is allowed to access.
- Line 17 calls `init()` to start stepping through the states of our DApp.

Now that our users are defined, the first step from our General Flow is to get the parameters from the `Admin`.
```
load: /examples/point-of-sale/index.rsh
md5:  2d9202924ee606ab47726038fbde05a3
range: 19-25
```
- Line 19 starts a local step.
- Line 20 gets the parameters from the `Admin` frontend and unpacks them into respective constants.
- Line 22 `{!rsh} publish`es these identifiers to the blockchain.
- Line 23 has us `{!rsh} commit()` to move out of consensus. Now our contract is aware of `tok` and it can be paid into the contract.
- Line 24 `{!rsh} pay`s the total supply of Loyalty Tokens into the contract.
- Line 25 Notifies the frontend that our contract is ready to start accepting `{!rsh} API` calls.

The first two steps of our General Flow are now complete.

> Now is a good time to get into the frontend `mjs` test suite and get that started.

Reviewer: ```suggestion
Now is a good time to get into the frontend (`.mjs`) test suite and get that started.
```

```
load: /examples/point-of-sale/index.mjs
md5:  5229e4cebe07397fe5e73a60dc19f70f
range: 1-15
```
- Lines 8-9 create a new account for our `Admin` and get the contract handle.
- Line 10 launches our Loyalty Token, depositing the total supply to `accA`.
- Lines 12-15 is a welcome message that is probably too long.

> Now let's automate our `Admin` interaction. Note the jump in line numbers, this code goes at the bottom of your `mjs` file.

Reviewer: ```suggestion
Now let's automate our `Admin` interaction. Please pay attention to the jump in line numbers: this code goes at the bottom of your `mjs` file.
```

```
load: /examples/point-of-sale/index.mjs
md5:  5229e4cebe07397fe5e73a60dc19f70f
range: 45-56
```
- Line 46-49 are various parameter values provided by the Admin.

> - Line 51-53 is our function that will be called when the contract is ready to start accepting `{!rsh} API` calls. Note the `startBuyers()` function that we will implement later.

Reviewer: ```suggestion
- Line 51-53 is our function that will be called when the contract is ready to start accepting `{!rsh} API` calls. We will implement the `startBuyers()` function later.
```
